File: 1530/solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    int uuid = 1;
    Set<Integer> leaves = new HashSet<>();
    Map<Integer, List<Integer>> graph = new HashMap<>();

    public int countPairs(TreeNode root, int distance) {
        if (root.left == null && root.right == null) {
            return 0;
        }
        root.val = 0;
        traverse(root);
        int res = 0;
        for (int leaf : leaves) {
            Deque<int[]> bfs = new ArrayDeque<>();
            Set<Integer> visited = new HashSet<>();
            bfs.add(new int[] { leaf, 0 });
            visited.add(leaf);
            while (!bfs.isEmpty()) {
                int[] curr = bfs.pollFirst();
                int node = curr[0];
                int dist = curr[1];
                if (dist == distance) {
                    continue;
                }
                for (int child : graph.get(node)) {
                    if (!visited.contains(child)) {
                        visited.add(child);
                        if (leaves.contains(child)) {
                            res++;
                        }
                        bfs.add(new int[] { child, dist + 1 });
                    }
                }
            }
        }
        return res / 2;
    }

    private void traverse(TreeNode node) {
        if (node.left != null) {
            node.left.val = uuid++;
            graph.computeIfAbsent(node.val, k -> new ArrayList<Integer>()).add(node.left.val);
            graph.computeIfAbsent(node.left.val, k -> new ArrayList<Integer>()).add(node.val);
            traverse(node.left);
        }
        if (node.right != null) {
            node.right.val = uuid++;
            graph.computeIfAbsent(node.val, k -> new ArrayList<Integer>()).add(node.right.val);
            graph.computeIfAbsent(node.right.val, k -> new ArrayList<Integer>()).add(node.val);
            traverse(node.right);
        }

        if (node.left == null && node.right == null) {
            leaves.add(node.val);
        }
    }
}

// time O(N * D^2)
// space O(N)

File: 1530/solution.ts
function countPairs(root: TreeNode | null, distance: number): number {
  const leaves = new Set<number>();
  const graph = new Map<number, number[]>();
  let uuid = 1;
  root!.val = 0;
  graph.set(0, []);

  const traverse = (node: TreeNode | null) => {
    if (!node) return;

    if (node.left) {
      node.left.val = uuid++;
      graph.set(node.left.val, []);
      graph.get(node.val)!.push(node.left.val);
      graph.get(node.left.val)!.push(node.val);
      traverse(node.left);
    }

    if (node.right) {
      node.right.val = uuid++;
      graph.set(node.right.val, []);
      graph.get(node.val)!.push(node.right.val);
      graph.get(node.right.val)!.push(node.val);
      traverse(node.right);
    }

    if (!node.left && !node.right) {
      leaves.add(node.val);
    }
  };

  traverse(root);

  let res = 0;
  for (const node of leaves) {
    const visited = new Set<number>();
    const bfs: [number, number][] = [[node, 0]];
    visited.add(node);

    while (bfs.length) {
      const [curr, dist] = bfs.shift()!;
      if (dist < distance) {
        for (const neighbor of graph.get(curr)!) {
          if (!visited.has(neighbor)) {
            visited.add(neighbor);
            if (leaves.has(neighbor)) {
              res++;
            }
            bfs.push([neighbor, dist + 1]);
          }
        }
      }
    }
  }
  return res / 2;
}
